### PR TITLE
fix for storing the correct object definition in catalog

### PR DESCRIPTION
**Pull request by pranavJ23**
### Description
Problem: If we create a procedure(let's say) and in that procedure we declare a variable of of some datatype with brackets around like 'DECLARE @SQL [varchar] (8000)' , the babelfish catalog stores it as 'DECLARE @SQL varchar (8000)'  i.e the brackets are not stored properly in catalog.
Or if we add a single line comment before a query(like alter, create query) then while logging the definition of that query from catalog we do not get the single line comment, which is not correct.

RCA: while parsing the query we mutate the parse tree in such a way that it modifies various things from the original string  like single line comment, brackets, multi line comments.

Fix: store the original query string before mutation and convert alter to create in case or alter procedure/function( batch level statement) and pass that query string. 
### Issues Resolved
[BABEL-5237][BABEL-5236][BABEL-5140]
### Sign Off
Signed-off-by: Pranav Jain [pranavjc@amazon.com](mailto:pranavjc@amazon.com)
### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: test/JDBC/expected/alter-function-vu-verify.out [hunk 1]
```diff
@@ -167,7 +167,7 @@ select ROUTINE_NAME, ROUTINE_BODY, ROUTINE_DEFINITION from information_schema.ro
 go
 ~~START~~
 nvarchar#!#nvarchar#!#nvarchar
-alter_func_f2#!#SQL#!#CREATE function alter_func_f2(@param2 varchar(10)) returns varchar(10) begin return @param2 end
+alter_func_f2#!#SQL#!#ALTER function alter_func_f2(@param2 varchar(10)) returns varchar(10) begin return @param2 end<newline>
```
Comment (tanscorpio7) [inline test/JDBC/expected/alter-function-vu-verify.out]: This does not look expected

**Review by tanscorpio7 — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/src/tsqlIface.cpp [hunk 2]
```diff
@@ -4016,7 +4015,26 @@ handleBatchLevelStatement(TSqlParser::Batch_level_statementContext *ctx, tsqlSel
...
 	// create PLtsql_stmt_execsql to wrap all query string
 	PLtsql_stmt_execsql *execsql = (PLtsql_stmt_execsql *) makeSQL(ctx);
-	execsql->original_query = pstrdup((makeTsqlExpr(ctx, false))->query);
+	if((ctx->create_or_alter_procedure() && ctx->create_or_alter_procedure()->ALTER()))
+	{
+		int startIndex = ctx->create_or_alter_procedure()->ALTER()->getSymbol()->getStartIndex();
+		std::string result = originalQuery;
+		int endIndex = startIndex + 5;
+		result.replace(startIndex, endIndex - startIndex, "CREATE");
+		execsql->original_query = pstrdup(result.c_str());
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/src/tsqlIface.cpp]: Lets not pstrdup here, instead pass a duplicate copy of the string in `handleBatchLevelStatement`.

File: contrib/babelfishpg_tsql/src/tsqlIface.cpp [hunk 3]
```diff
@@ -4016,7 +4015,26 @@ handleBatchLevelStatement(TSqlParser::Batch_level_statementContext *ctx, tsqlSel
...
+	}
+	else if (ctx->create_or_alter_function() && ctx->create_or_alter_function()->ALTER())
+	{
+		int startIndex = ctx->create_or_alter_function()->ALTER()->getSymbol()->getStartIndex();
+		std::string result = originalQuery;
+		int endIndex = startIndex + 5;
+		result.replace(startIndex, endIndex - startIndex, "CREATE");
+		execsql->original_query = pstrdup(result.c_str());
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/src/tsqlIface.cpp]: We can put this code into another function to avoid duplicate code

**Review by tanscorpio7 — CHANGES_REQUESTED** [review]

File: test/JDBC/expected/BABEL_OBJECT_DEFINITION_before_16_5-vu-prepare.out [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+GO
+
+-- Inline table-valued function
+CREATE FUNCTION object_definition_itvf()
+RETURNS table
+AS
+RETURN (SELECT 42 AS VALUE)
+go
+
+-- Table-valued function
```
Comment (shalinilohia50) [inline test/JDBC/expected/BABEL_OBJECT_DEFINITION_before_16_5-vu-prepare.out:44]: Please add test for multi line comments.

File: contrib/babelfishpg_tsql/src/tsqlIface.cpp [hunk 5]
```diff
@@ -3992,8 +3992,33 @@ get_start_token_of_batch_level_stmt_body(TSqlParser::Batch_level_statementContex
 	return nullptr;
 }
 
+char*
+createHandlingForBatchLevelStatement(TSqlParser::Batch_level_statementContext *ctx, const char *originalQuery)
```
Comment (shalinilohia50) [inline contrib/babelfishpg_tsql/src/tsqlIface.cpp]: Please rename it to something meaningful like `storeOriginalQueryForBatchLevelStatement`.

File: test/JDBC/expected/BABEL_OBJECT_DEFINITION_before_16_5-vu-prepare.out [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+GO
+
+-- Procedure
+CREATE PROC object_definition_proc
+AS
+SELECT 1;
+GO
+
+-- Scalar function
+CREATE FUNCTION object_definition_fc1(@fc1_a nvarchar) RETURNS nvarchar AS BEGIN return @fc1_a END;
```
Comment (shalinilohia50) [inline test/JDBC/expected/BABEL_OBJECT_DEFINITION_before_16_5-vu-prepare.out:26]: Let's add test where the datatype is within the brackets like `[varchar]`.

**Review by tanscorpio7 — APPROVED** [review]
Please add all the test cases mentioned in all three JIRAs to JDBC & DDL framework.

**Comment by shalinilohia50** [comment]
1. Let's add tests in DDL framework as well for single/multi line comments and brackets around the datatype.
2. Please don't create another test file(storeOriginalQuery*) just to show function object definition, rather use the existing ones(BABEL_OBJECT_DEFINITION*).

**Review by tanscorpio7 — APPROVED** [review]

**Review by shalinilohia50 — APPROVED** [review]

**Review by shalinilohia50 — APPROVED** [review]

**Review by tanscorpio7 — APPROVED** [review]

File: contrib/babelfishpg_tsql/src/tsqlIface.cpp [hunk 7]
```diff
@@ -3992,8 +3992,33 @@ get_start_token_of_batch_level_stmt_body(TSqlParser::Batch_level_statementContex
...
+	int startIndex = -1;
+	int endIndex = -1;
+	std::string originalQueryCopy = originalQuery;
+
+	if ((ctx->create_or_alter_procedure() && ctx->create_or_alter_procedure()->ALTER()))
+	{
+		startIndex = ctx->create_or_alter_procedure()->ALTER()->getSymbol()->getStartIndex();
+		endIndex = startIndex + 5;
+		originalQueryCopy.replace(startIndex, endIndex - startIndex, "CREATE");
+		return pstrdup(originalQueryCopy.c_str());
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/src/tsqlIface.cpp:4007]: nit:
We can refactor it by moving following 3 lines out of if ... else.
```
endIndex = startIndex + 5;
originalQueryCopy.replace(startIndex, endIndex - startIndex, "CREATE");
return pstrdup(originalQueryCopy.c_str());
```

**Review by HarshLunagariya — APPROVED** [review]

**Review by surendravishnoi — DISMISSED** [review]

**Review by shardgupta — APPROVED** [review]

**Pull request merged**